applications/GuermondNeumannBC.cc

```diff
@@ -173,19 +175,35 @@ void Guermond<dim>::setup_constraints()
 
   velocity_exact_solution->set_time(time_stepping.get_start_time());
 
+  // left boundary
   velocity->boundary_conditions.set_neumann_bcs(0);
+  // right boundary
   velocity->boundary_conditions.set_dirichlet_bcs(
     1,
     velocity_exact_solution,
     true);
+  pressure->boundary_conditions.set_dirichlet_bcs(
+  	1,
+		pressure_exact_solution,
+		true);
+  // bottom boundary
   velocity->boundary_conditions.set_dirichlet_bcs(
     2,
     velocity_exact_solution,
     true);
+  pressure->boundary_conditions.set_dirichlet_bcs(
+  	2,
+		pressure_exact_solution,
+		true);
+  // top boundary
   velocity->boundary_conditions.set_dirichlet_bcs(
     3,
     velocity_exact_solution,
     true);
+  pressure->boundary_conditions.set_dirichlet_bcs(
+  	3,
+		pressure_exact_solution,
+		true);
```

Comment:
This is the problem. By enforcing this boundary conditions on the pressure you are formulating another problem in the `phi` space due to how I constrain `phi` inside `NavierStokesProjection` according to the pressure and velocity spaces. The original Neumann problem translate into
```
+---------+------------+------------------------------------------+
| Boundary conditions of the Velocity entity                      |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| 0       | Neumann    | ZeroTensorFunction<1, 2, double>         |
| 1       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
| 2       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
| 3       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
+---------+------------+------------------------------------------+
+---------+------------+------------------------------------------+
| Boundary conditions of the Pressure entity                      |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| Unconstrained boundary ids: 0, 1, 2, 3                          |
+---------+------------+------------------------------------------+
+---------+------------+------------------------------------------+
| Boundary conditions of the Phi entity                           |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| 1       | Neumann    | ZeroFunction<2, double>                  |
| 2       | Neumann    | ZeroFunction<2, double>                  |
| 3       | Neumann    | ZeroFunction<2, double>                  |
| 0       | Dirichlet  | ZeroFunction<2, double>                  |
+---------+------------+------------------------------------------+
```
but with your changes the problem becomes
```
+---------+------------+------------------------------------------+
| Boundary conditions of the Velocity entity                      |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| 0       | Neumann    | ZeroTensorFunction<1, 2, double>         |
| 1       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
| 2       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
| 3       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
+---------+------------+------------------------------------------+
+---------+------------+------------------------------------------+
| Boundary conditions of the Pressure entity                      |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| 1       | Dirichlet  | GuermondNeumannBC::PressureExactSolution<2> |
| 2       | Dirichlet  | GuermondNeumannBC::PressureExactSolution<2> |
| 3       | Dirichlet  | GuermondNeumannBC::PressureExactSolution<2> |
| Unconstrained boundary ids: 0                                   |
+---------+------------+------------------------------------------+
+---------+------------+------------------------------------------+
| Boundary conditions of the Phi entity                           |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| 0       | Dirichlet  | ZeroFunction<2, double>                  |
| 1       | Dirichlet  | ZeroFunction<2, double>                  |
| 2       | Dirichlet  | ZeroFunction<2, double>                  |
| 3       | Dirichlet  | ZeroFunction<2, double>                  |
+---------+------------+------------------------------------------+
```

Reply:
Okay I understand. With the original code I get 
```
+---------+------------+------------------------------------------+
| Boundary conditions of the Velocity entity                      |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| 0       | Neumann    | ZeroTensorFunction<1, 2, double>         |
| 1       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
| 2       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
| 3       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
+---------+------------+------------------------------------------+
+---------+------------+------------------------------------------+
| Boundary conditions of the Pressure entity                      |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| Unconstrained boundary ids: 0, 1, 2, 3                          |
+---------+------------+------------------------------------------+
```

but the problem is that I get the runtime error

```
----------------------------------------------------
Exception on processing: 

--------------------------------------------------------
An error occurred in line <624> of file </home/sg/Documents/Projects/Git/RotatingMHD/source/entities_structs.cc> in function
    void RMHD::Entities::ScalarEntity<dim>::apply_boundary_conditions() [with int dim = 2]
The violated condition was: 
    boundary_conditions.regularity_guaranteed()
Additional information: 
    No boundary conditions were set for the "Pressure" entity
--------------------------------------------------------

Aborting!
----------------------------------------------------
```
because the pressure does not have any boundary condition. Should I apply a mean value constraint or should we resolve this?

Reply:
> because the pressure does not have any boundary condition. Should I apply a mean value constraint or should we resolve this?

Applying a mean value constraint would formulate a different problem. We have to build a bypass of the regularity check for the pressure. As it's regularity can be ensured indirectly. A not so elegant solution is just to pass a flag to the constructor to indicate the bypass. A general scalar field, i.e. the temperature field, needs this regularity check so deleting is not really a solution.